src/Databases/DatabaseReplicated.cpp

```diff
@@ -0,0 +1,288 @@
+
+        // The third condition gurantees at most one snapshot creation per batch
+        if (log_n > 0 && snapshot_period > 0 && (last_log_n - log_n) / snapshot_period == 0 && log_n % snapshot_period == 0) {
+            createSnapshot();
+        }
+    }
+
+    background_log_executor->scheduleAfter(500);
+}
+
+void DatabaseReplicated::writeLastExecutedToDiskAndZK() {
+    current_zookeeper = getZooKeeper();
+    current_zookeeper->createOrUpdate(zookeeper_path + "/replicas/" + replica_name, last_executed_log_entry, zkutil::CreateMode::Persistent);
+
+    String metadata_file = getMetadataPath() + ".last_entry";
+    WriteBufferFromFile out(metadata_file, last_executed_log_entry.size(), O_WRONLY | O_CREAT);
+    writeString(last_executed_log_entry, out);
+    out.next();
+    if (global_context.getSettingsRef().fsync_metadata)
+        out.sync();
+    out.close();
+}
+
+void DatabaseReplicated::executeFromZK(String & path) {
+        current_zookeeper = getZooKeeper();
+        String query_to_execute = current_zookeeper->get(path, {}, NULL);
+        ReadBufferFromString istr(query_to_execute);
+        String dummy_string;
+        WriteBufferFromString ostr(dummy_string);
+
+        try
+        {
+            current_context = std::make_unique<Context>(global_context);
+            current_context->getClientInfo().query_kind = ClientInfo::QueryKind::REPLICATED_LOG_QUERY;
+            current_context->setCurrentDatabase(database_name);
+            current_context->setCurrentQueryId(""); // generate random query_id
+            executeQuery(istr, ostr, false, *current_context, {});
+        }
+        catch (...)
+        {
+            tryLogCurrentException(log, "Query from zookeeper " + query_to_execute + " wasn't finished successfully");
+    
+        }
+
+        LOG_DEBUG(log, "Executed query: " << query_to_execute);
+}
+
+void DatabaseReplicated::propose(const ASTPtr & query) {
+    current_zookeeper = getZooKeeper();
+
+    LOG_DEBUG(log, "Proposing query: " << queryToString(query));
+    current_zookeeper->create(zookeeper_path + "/log/log-", queryToString(query), zkutil::CreateMode::PersistentSequential);
+
+    background_log_executor->schedule();
+}
+
+void DatabaseReplicated::createSnapshot() {
+    current_zookeeper = getZooKeeper();
+    String snapshot_path = zookeeper_path + "/snapshots/" + last_executed_log_entry;
+
+    if (Coordination::ZNODEEXISTS == current_zookeeper->tryCreate(snapshot_path, String(), zkutil::CreateMode::Persistent)) {
+        return;
+    }
+    
+    for (auto iterator = getTablesIterator({}); iterator->isValid(); iterator->next()) {
+        String table_name = iterator->name();
+        auto query = getCreateQueryFromMetadata(getObjectMetadataPath(table_name), true);
+        String statement = queryToString(query);
+        current_zookeeper->createOrUpdate(snapshot_path + "/" + table_name, statement, zkutil::CreateMode::Persistent);
```

Comment:
Why `OrUpdate`?

Reply:
No particular reason.
It's mostly `set` or `trySet` and when the replica is initialized, it's a `create`. `createOrUpdate` does exactly this.

Reply:
`createIfNotExists` is more suitable here

Reply:
Then is it possible, that such zknodes are already exist since each `snapshot_path` can be created only once and only by one replica?

Reply:
Yeah, you're right. Due to the check `Coordination::ZNODEEXISTS == current_zookeeper->tryCreate(snaphot_path)` there's no way several replicas would create the same snapshot.